webclient/doi/model_serializers.py

```diff
@@ -0,0 +1,994 @@
+            except ObjectDoesNotExist:
+                pass
+
+            try:
+                if cast.archeologicky_zaznam and cast.archeologicky_zaznam.stav == AZ_STAV_ARCHIVOVANY:
+                    serialized_subjects.append(serialize_subject(cast.archeologicky_zaznam.lokalita.typ_lokality))
+                    serialized_subjects.append(serialize_subject(cast.archeologicky_zaznam.lokalita.druh))
+            except ObjectDoesNotExist:
+                pass
+        serialized_subjects = [dict(item) for item in set(serialized_subjects) if item]
+        return serialized_subjects
+
+    def _serialize_types(self) -> dict:
+        resource_type_query = self.record.typ_dokumentu.heslar_odkaz.filter(zdroj="DataCite").filter(
+            nazev_kodu="resourceTypeGeneral"
+        )
+        serialized_types = {}
+        if resource_type_query.exists():
+            serialized_types["resourceType"] = resource_type_query.first().heslo_en
+            serialized_types["resourceTypeGeneral"] = resource_type_query.first().kod
+        else:
+            serialized_types["resourceType"] = "Dataset"
+            serialized_types["resourceTypeGeneral"] = "Dataset"
+        return serialized_types
+
+    def serialize_hide(self):
+        return {
+            "data": {
+                "type": "dois",
+                "attributes": {
+                    "event": "hide",
+                    "descriptions": self._serialize_descriptions()
+                    + [
+                        {
+                            "lang": "cs",
+                            "description": self.DELETE_DESCRIPTION_CZ,
+                            "descriptionType": "TechnicalInfo",
+                        },
+                        {
+                            "lang": "en",
+                            "description": self.DELETE_DESCRIPTION_EN,
+                            "descriptionType": "TechnicalInfo",
+                        },
+                    ],
+                },
+            }
+        }
+
+
+class SamostatnyNalezSerializer(ModelSerializer):
+    def __init__(self, record: SamostatnyNalez):
+        super().__init__(record)
+        self.record: SamostatnyNalez
+
+    def _get_creators(self):
+        return [self.record.nalezce]
+
+    def _get_historie_queryset(self):
+        return self.record.historie.historie_set
+
+    def get_ident_cely(self):
+        return self.record.ident_cely
+
+    def _get_soubory_queryset(self):
+        if self.record.soubory:
+            return self.record.soubory.soubory
+
+    def _get_prefix(self):
+        return settings.IGSN_PREFIX
+
+    def _get_publication_year(self):
+        publication_year_history = self._get_historie_queryset().filter(typ_zmeny=ARCHIVACE_SN)
+        if publication_year_history.exists():
+            return publication_year_history.first().datum_zmeny.year
+        else:
+            # DataCite request may be called before the history record is created
```

Comment:
Je dobře, že se s tím počítá, ale kdy ses s tím v praxi setkal, že to dělalo problém? Je to myslím spíš potenciální chyba a mělo by to vrátit výjimku.

Reply:
Ono se to stane vždy při prvotní archivaci, protože DOI se generuje jako první (viz třeba zde: https://github.com/ARUP-CAS/aiscr-webamcr/blob/e627e5131c089bb54c1264ac61d1fdc35849e029/webclient/pas/views.py#L599).

Reply:
Aha, už je mi to jasné.